src/common/types.ts

```diff
@@ -160,12 +161,246 @@ export type CSVRecord<Header extends ReadonlyArray<string>> = Record<
   string
 >;
 
+type Concat<T extends any[]> = T extends [infer F, ...infer R]
+  ? R extends any[]
+    ? F & Concat<R>
+    : F
+  : // biome-ignore lint/complexity/noBannedTypes: <explanation>
+    {};
+
+type Replace<
+  Target extends string,
+  From extends string,
+  To extends string,
+> = Target extends `${infer A}${From}${infer B}`
+  ? Replace<`${A}${To}${B}`, From, To>
+  : Target;
+
+type Split<
+  Char extends string,
+  Delimiter extends string = typeof COMMA,
+> = Char extends `${infer F}${Delimiter}${infer R}`
+  ? [F, ...Split<R, Delimiter>]
+  : [Char];
+
+type Join<
+  Chars extends ReadonlyArray<string | number | boolean | bigint>,
+  Delimiter extends string = typeof COMMA,
+  Quotation extends string = typeof DOUBLE_QUOTE,
+> = Chars extends readonly [infer F, ...infer R]
+  ? F extends string
+    ? R extends string[]
+      ? `${F extends `${string}${Newline | Delimiter}${string}`
+          ? `${Quotation}${F}${Quotation}`
+          : F}${R extends [] ? "" : Delimiter}${Join<R, Delimiter, Quotation>}`
+      : string
+    : string
+  : "";
+
+type Newline = typeof CR | typeof CRLF | typeof LF;
+type DummyNewline<NL extends Newline> =
+  `web-csv-toolbox.DummyNewline-${NL extends typeof CRLF
+    ? "crlf"
+    : NL extends typeof CR
+      ? "cr"
+      : "lf"}`;
+type DummyDelimiter = "web-csv-toolbox.DummyDelimiter";
+
+type SplitNewline<T extends string> =
+  T extends `${infer A}${typeof CRLF}${infer B}`
+    ? [...Split<A, typeof CRLF>, ...SplitNewline<B>]
+    : T extends `${infer A}${typeof CR}${infer B}`
+      ? [...Split<A, typeof CR>, ...SplitNewline<B>]
+      : T extends `${infer A}${typeof LF}${infer B}`
+        ? [...Split<A, typeof LF>, ...SplitNewline<B>]
+        : [T];
+
+type Newline2DummyNewline<T extends string> =
+  T extends `${infer A}${typeof CRLF}${infer B}`
+    ? Newline2DummyNewline<`${A}${DummyNewline<typeof CRLF>}${B}`>
+    : T extends `${infer A}${typeof CR}${infer B}`
+      ? Newline2DummyNewline<`${A}${DummyNewline<typeof CR>}${B}`>
+      : T extends `${infer A}${typeof LF}${infer B}`
+        ? Newline2DummyNewline<`${A}${DummyNewline<typeof LF>}${B}`>
+        : T;
+
+type DummyNewline2Newline<T extends string> =
+  T extends `${infer A}${DummyNewline<typeof CRLF>}${infer B}`
+    ? DummyNewline2Newline<`${A}${typeof CRLF}${B}`>
+    : T extends `${infer A}${DummyNewline<typeof CR>}${infer B}`
+      ? DummyNewline2Newline<`${A}${typeof CR}${B}`>
+      : T extends `${infer A}${DummyNewline<typeof LF>}${infer B}`
+        ? DummyNewline2Newline<`${A}${typeof LF}${B}`>
+        : T;
+
+type Escape<
```

Comment:
I noticed that the current implementation of the Escape type might not correctly handle cases where quotation marks are escaped by doubling them within CSV data. This is a common scenario in CSV formats where quotation marks within a field are doubled to denote that they are literal characters and not delimiters.

For example, consider the CSV line:

```csv
"Smith, John","He said, ""This is awesome!"""
```

In this line, the inner quotation marks around "This is awesome!" are doubled to escape them, but our current escaping logic may not preserve these as expected. This could lead to incorrect parsing or data corruption when the CSV is processed.

To address this, I recommend enhancing our escape handling logic to detect and properly preserve doubled quotation marks within quoted fields. This would ensure that our CSV processing remains robust and accurate, particularly for data that includes literal quotation marks.